### PR TITLE
Container Setup Test Rewrite With Mocks

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

The _ContainerSetupSuite_ tests previously used _JujuConnSuite_ and were racy.

This patch rectifies this with the following changes:
- _JujuConnSuite_ is no longer used. All dependencies are mocked.
- Some tests were removed that were tautological or that effectively tested functionality from other modules.
- Some mocks that were previously located close to their usage have been moved to more general locations, which necessitates some edits to code no directly related to container setup.
- Errors traces have been added to places where they were previously absent in the course of debugging.

_Note for Reviewers_: worker/provisioner/container_initialisation_test.go bears little resemblance to its old form. It might be better than to read this in its own right rather than the diff.

## QA steps

The re-written tests pass consistently.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Rewrite Container Provisioner Tests](https://trello.com/c/4QwTAZlD/607-rewrite-container-provisioner-tests)

File: container/factory/factory.go [hunk 1]
```diff
@@ -16,7 +16,7 @@ import (
 
 // NewContainerManager creates the appropriate container.Manager for the
 // specified container type.
-func NewContainerManager(forType instance.ContainerType, conf container.ManagerConfig) (container.Manager, error) {
+var NewContainerManager = func(forType instance.ContainerType, conf container.ManagerConfig) (container.Manager, error) {
```
Comment (SimonRichardson) [inline container/factory/factory.go]: Is this needed? I'm not sure why you would want a mutable `NewContainerManager`, unless you want to mock it out for tests? It's not obviously used else where in this PR.

[the remark this answers is not in the document]
Comment (manadart) [reply container/factory/factory.go]: Agreed. I don't like calls to patch at all.

If we lived in an ideal world, all these factory-type dependencies would be injected at the top, through the manifold by the dependency engine. Then at worker creation you could pass in a mock by its indirection.

Rolling it all back up would be a hell of a job though.

**Comment by manadart** [comment]
Check job looks like an intermittent failure. Let's try a $$merge$$.

**Pull request merged**